[TextFields] Correct UIAccessibility values and behavior.

romoore opened the pull request:
## `accessibilityLabel` is used for all non-user-input values in `MDCTextField`.
Previously, this text was all presented in the `accessibilityValue` of the text field, which resulted in weird effects, including:
*  Changes in pitch to indicate that the value had changed (up to indicate additions, down for deletions).
*  The entire error message being read out as if it were the value.
*  The placeholder and labels being read out as if they were the value.

With this change, anything not input by the user (or programmatically set as if the user had done so) is now moved to `accessibilityLabel`.  This more closely matches the intent of the property.  Specifically, the `accessibilityLabel` property will return the contents of the following (if present):
1.   The placeholder text.
2.  The leading underline label.
3.  The trailing underline label. **This was previously excluded from the `accessibilityValue`, though it's not clear why.**

The `accessibilityLabel` will also introduce commas (slight pauses) between each of these values when concatenating them into a single string.  This alleviates some of the difficulty in mentally separating each of the values when hearing it read by VoiceOver.

## `UIAccessibilityAnnouncementNotification` is now used to announce the error to the user in `MDCTextInputControllerBase`.
Previously, a `UIAccessibilityLayoutChangedNotification` was used and pointed to the `leadingUnderlineLabel`.  However, this could easily result in the user becoming "trapped" in the label because it was an accessibility element outside of a "container" element. This is incorrect behavior - any UIView that returns `isAccessibilityElement = YES` should summarize its subviews.

Because announcements can be interrupted, and because announcements are already used for things like character counters, *error messages can interrupt other announcements*. This is intentional, however, since errors should be corrected immediately and therefore it's reasonable to cut short any other announcements generated by the text field.  A slight delay (not currently customizable) is introduced to allow an opportunity for the text field to read-out any content that was added or removed just before the error is announced.

Fixes #7157

Commit:
[TextFields] Fix error/helper text with VoiceOver.

When the error or helper text is changed and an announcement is made to draw
the focus of UIAccessibility to the label, a user can be "trapped" in the
label without a mechanism to move to other UI elements.

Fixes #7157

Commit:
Merge branch 'develop' into issue-7157

Commit:
Clean up dead code.

Commit:
More cleanup

Commit:
Revert work

Commit:
Remove other stuff

material-automation commented:
bazel detected changes to the following targets:

```
//components/Chips:snapshot_tests
//components/Chips:snapshot_tests_test_bundle
//components/Chips:unit_tests
//components/Chips:unit_tests_IPAD_PRO_12_9_IN_9_3
//components/Chips:unit_tests_IPAD_PRO_12_9_IN_9_3_test_bundle
//components/Chips:unit_tests_IPHONE_7_PLUS_IN_10_3
//components/Chips:unit_tests_IPHONE_7_PLUS_IN_10_3_test_bundle
//components/Chips:unit_tests_IPHONE_X_IN_11_0
//components/Chips:unit_tests_IPHONE_X_IN_11_0_test_bundle
//components/TextFields:snapshot_tests
//components/TextFields:snapshot_tests_test_bundle
//components/TextFields:unit_tests
//components/TextFields:unit_tests_IPAD_PRO_12_9_IN_9_3
//components/TextFields:unit_tests_IPAD_PRO_12_9_IN_9_3_test_bundle
//components/TextFields:unit_tests_IPHONE_7_PLUS_IN_10_3
//components/TextFields:unit_tests_IPHONE_7_PLUS_IN_10_3_test_bundle
//components/TextFields:unit_tests_IPHONE_X_IN_11_0
//components/TextFields:unit_tests_IPHONE_X_IN_11_0_test_bundle
```

<!-- identifier: affected-targets -->

material-automation commented on components/TextFields/src/MDCTextInputControllerBase.m:
```diff
@@ -1578,18 +1575,10 @@ - (void)setErrorText:(NSString *)errorText
-    }
+    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.1 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
+      UIAccessibilityPostNotification(UIAccessibilityAnnouncementNotification, errorText);
+    });
```
clang-format suggested the following change:


```
-    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.1 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
-      UIAccessibilityPostNotification(UIAccessibilityAnnouncementNotification, errorText);
-    });
+    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(0.1 * NSEC_PER_SEC)),
+                   dispatch_get_main_queue(), ^{
+                     UIAccessibilityPostNotification(UIAccessibilityAnnouncementNotification,
+                                                     errorText);
+                   });
```

<!-- identifier: clang-format -->
<!-- File: components/TextFields/src/MDCTextInputControllerBase.m -->

material-automation commented on components/TextFields/src/MDCTextField.m:
```diff
@@ -792,18 +792,34 @@ - (void)mdc_setAdjustsFontForContentSizeCategory:(BOOL)adjusts {
   }
   return accessibilityStrings.count > 0 ? [accessibilityStrings componentsJoinedByString:@" "]
-                                        : nil;
+  : nil;
```
clang-format suggested the following change:


```
-  : nil;
+                                        : nil;
```

<!-- identifier: clang-format -->
<!-- File: components/TextFields/src/MDCTextField.m -->

material-automation commented:
clang-format recommended changes.

Consider installing [git-clang-format](https://github.com/llvm-mirror/clang/blob/master/tools/clang-format/git-clang-format) and running the following command:

```bash
git clang-format $(git merge-base origin/develop HEAD)
```

<!-- identifier: clang-format -->

Commit:
Merge branch 'develop' into issue-7157

Commit:
[TextFields] Make Kitchen Sink better for tests.

Adding "validation" to make it easier to test error states in the
Kitchen Sink example.

Preparation for #7157

Commit:
Make Kitchen Sink easier to test with

Commit:
Fix delay, add trailing label.

Commit:
Minor clean-up of `accessibilityValue`

material-automation commented on components/TextFields/src/MDCTextInputControllerBase.m:
```diff
@@ -1581,18 +1579,10 @@ - (void)setErrorText:(NSString *)errorText
-    }
+    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(kDefaultErrorAnnouncementDelay * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
+      UIAccessibilityPostNotification(UIAccessibilityAnnouncementNotification, errorText);
+    });
```
clang-format suggested the following change:


```
-    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(kDefaultErrorAnnouncementDelay * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
-      UIAccessibilityPostNotification(UIAccessibilityAnnouncementNotification, errorText);
-    });
+    dispatch_after(
+        dispatch_time(DISPATCH_TIME_NOW, (int64_t)(kDefaultErrorAnnouncementDelay * NSEC_PER_SEC)),
+        dispatch_get_main_queue(), ^{
+          UIAccessibilityPostNotification(UIAccessibilityAnnouncementNotification, errorText);
+        });
```

<!-- identifier: clang-format -->
<!-- File: components/TextFields/src/MDCTextInputControllerBase.m -->

Commit:
Revert "Make Kitchen Sink easier to test with"

This reverts commit a19b0023b18993f657014ff1f4a8630a8a78662e.

Commit:
Formatting

Commit:
Make A11Y value different from displayed.

This makes it easier to ensure we're using the a11y value and not the
displayed value when presenting information to VoiceOver.

Commit:
Restore accessibility values to labels

Commit:
Revert "Make A11Y value different from displayed."

This reverts commit 9c09897eeebb5092a8dc8ae9b638d46ae6cd384a.

Commit:
Revert "[TextFields] Make Kitchen Sink better for tests."

This reverts commit 7aaa065793f10946a01f9b3b24f9d16be8f1701b.

material-automation commented on components/TextFields/src/MDCTextInputControllerBase.m:
```diff
@@ -1570,29 +1571,22 @@ - (void)setErrorText:(NSString *)errorText
     self.textInput.leadingUnderlineLabel.accessibilityLabel = announcementString;
-  } else {
-    self.textInput.accessibilityValue = nil;
+    
```
clang-format suggested the following change:


```
-    
+
```

<!-- identifier: clang-format -->
<!-- File: components/TextFields/src/MDCTextInputControllerBase.m -->

material-automation commented:
clang-format recommended changes.

Consider installing [git-clang-format](https://github.com/llvm-mirror/clang/blob/master/tools/clang-format/git-clang-format) and running the following command:

```bash
git clang-format $(git merge-base origin/develop HEAD)
```

<!-- identifier: clang-format -->

Commit:
Merge branch 'develop' into issue-7157

Commit:
Fix announcement string

romoore commented:
## IMPORTANT: IF YOU REVIEW

Please verify the new behavior **on a device using VoiceOver**.  Please do not grant LGTM/Approval until you have performed manual testing. I recommend using the Kitchen Sink or "Outlined Controller" examples.  Specifically pay attention to:

*  Error states - be sure the accessibility values are read-out.
*  Character counters and how they're read-out in different text field states.
*  Helper text and how it's read-out in different states.

codeman7 reviewed (dismissed):
Still want to test this on a device when I get in tomorrow morning.

codeman7 commented on components/TextFields/src/MDCTextField.m:
```diff
@@ -792,20 +792,39 @@ - (void)mdc_setAdjustsFontForContentSizeCategory:(BOOL)adjusts {
-- (NSString *)accessibilityValue {
+/*
+ Returns a combination of the following:
+ -  The superclass `accessibilityLabel` value
```
_Question:_ What was the motivation behind using \`accessibilityLabel\` instead of @c accessibilityLabel?

romoore replied:
In non-doc comments? I think it's just easier to read since we're mostly familiar with Markdown.

romoore commented:
@codeman7 Please feel free to merge this when you finish testing. I'm taking the label off, however, since I don't want a Good Samaritan passing by to merge this by accident.

codeman7 commented on components/TextFields/src/MDCTextInputControllerBase.m:
```diff
@@ -41,6 +41,7 @@
     MDCTextInputControllerBaseDefaultFloatingPlaceholderDownAnimationDuration = (CGFloat)0.266666;
 static const NSTimeInterval
     MDCTextInputControllerBaseDefaultFloatingPlaceholderUpAnimationDuration = (CGFloat)0.3;
+static const NSTimeInterval kDefaultErrorAnnouncementDelay = (CGFloat)0.500;
```
_Question:_ Where is this value coming from?

romoore replied:
I just made it up. It seems sufficient to vocalize most text input. It slightly mitigates the error announcement interruption effect.

Commit:
Merge branch 'develop' into issue-7157
